mysteriousorganism.js

```diff
@@ -0,0 +1,136 @@
+// Returns a random DNA base
+const returnRandBase = () => {
+	const dnaBases = ['A', 'T', 'C', 'G'];
+	return dnaBases[Math.floor(Math.random() * 4)];
+};
+
+// Returns a random single stand of DNA containing 15 bases
+const mockUpStrand = () => {
+	const newStrand = [];
+	for (let i = 0; i < 15; i++) {
+		newStrand.push(returnRandBase());
+	}
+	return newStrand;
+};
+
+// Create function factory
+const pAequorFactory = (specimenNum, dna) => {
+	return {
+		specimenNum,
+		dna,
+
+		mutate() {
+			const dnaBases = ['A', 'T', 'C', 'G']; // You'd probably want this variable to be global at some point.
+			const randomBase = Math.floor(Math.random() * 15);
+			const replaceBase = dna[randomBase];
+			let newBase = '';
+
+			// Find value at random place in base and replace it with a different base
+			do {
+				newBase = dnaBases[Math.floor(Math.random() * 3)];
+			} while (replaceBase === newBase);
+			dna.splice(randomBase, 1, newBase);
+			return dna;
+		},
+
+		compareDNA(inputObject) {
+			let compareCounter = 0;
+			for (i = 0; i < dna.length; i++) {
+				if (dna[i] === inputObject.dna[i]) {
+					compareCounter += 1;
+				}
+			}
+			return Math.floor((compareCounter / dna.length) * 100);
+		},
+
+		willLikelySurvive() {
+			let surviveCounter = 0;
+			for (i = 0; i < dna.length; i++) {
+				if (dna[i] === 'C' || dna[i] === 'G') {
+					surviveCounter += 1;
+				}
+			}
+			if (surviveCounter >= 9) {
+				return true;
+			} else return false;
+		},
+
+		complementStrand() {
+			let complementingStrand = [];
+			for (i = 0; i < dna.length; i++) {
+				if (dna[i] === 'A') {
+					complementingStrand.push('T');
+				}
+				if (dna[i] === 'T') {
+					complementingStrand.push('A');
+				}
+				if (dna[i] === 'C') {
+					complementingStrand.push('G');
+				}
+				if (dna[i] === 'G') {
+					complementingStrand.push('C');
+				}
+			}
+			return complementingStrand;
+		},
+	};
+};
+
+const createDatabase = factoryAmount => {
+	const strandArray = [];
+	let strandID = 1;
+	// if j = i then bug...
+	for (j = 0; j < factoryAmount; j++) {
+		let newStrand = pAequorFactory(strandID, mockUpStrand());
+		do {
+			newStrand.mutate();
+		} while (newStrand.willLikelySurvive() === false);
```

Comment:
Dit is wel interessant. Je kiest hier voor om een specimen net zo lang te muteren tot deze likely to survive is. Dat werkt uiteraard, want er is OOIT wel een moment dat het specimen dusdanig ver gemuteerd is dat het kan overleven. Ik vraag me wel af of het minder operaties nodig heeft dan telkens een nieuw specimen random laten maken. Waarschijnlijk is een hybride oplossing het aller snelst. Voor deze opdracht niet echt heel belangrijk, maar wel leuk om over te hebben eventueel.

Reply:
Oke, dit heb ik puur uit interesse even bekeken. Wat je doet werkt uiteraard. Maar je lijkt minder operators nodig te hebben als je een heel nieuw setje DNA maakt
![mysteriousorganismStefan_js_—_codecademy](https://user-images.githubusercontent.com/22873490/135728438-2244b148-b595-4122-9ac9-cb67cf7a66c9.jpg)
.

Reply:
Ja hier twijfelde ik over, een method gebruiken van het al bestaande object of een hele nieuwe DNA string maken. Eerste is dus beter als we het over snelheid/resources hebben.

Reply:
Ja nieuwe DNA is beter. Zie mijn file voor een leuke vergelijking.